src/Microsoft.ML.Transforms/CustomMappingCatalog.cs

```diff
@@ -46,5 +46,64 @@ public static class CustomMappingCatalog
             where TSrc : class, new()
             where TDst : class, new()
             => new CustomMappingEstimator<TSrc, TDst>(catalog.GetEnvironment(), mapAction, contractName, inputSchemaDefinition, outputSchemaDefinition);
+
+        /// <summary>
+        /// Create a <see cref="StatefulCustomMappingEstimator{TSrc, TState, TDst}"/>, which applies a custom mapping of input columns to output columns,
+        /// while allowing a per-cursor state.
+        /// </summary>
+        /// <typeparam name="TSrc">The class defining which columns to take from the incoming data.</typeparam>
+        /// <typeparam name="TState">The type that describes per-cursor state.</typeparam>
+        /// <typeparam name="TDst">The class defining which new columns are added to the data.</typeparam>
+        /// <param name="catalog">The transform catalog.</param>
+        /// <param name="mapAction">The mapping action. In addition to the input and output objects, the action is given a state object that it can look at and/or modify.
+        /// If the resulting transformer needs to be save-able, the class defining <paramref name="mapAction"/> should implement
+        /// <see cref="StatefulCustomMappingFactory{TSrc, TDst, TState}"/> and needs to be decorated with
+        /// <see cref="CustomMappingFactoryAttributeAttribute"/> with the provided <paramref name="contractName"/>.
+        /// The assembly containing the class should be registered in the environment where it is loaded back
+        /// using <see cref="ComponentCatalog.RegisterAssembly(System.Reflection.Assembly, bool)"/>.</param>
+        /// <param name="stateInitAction">The action to initialize the state object, that is called once before the cursor is initialized.</param>
+        /// <param name="contractName">The contract name, used by ML.NET for loading the model.
+        /// If <see langword="null"/> is specified, resulting transformer would not be save-able.</param>
+        /// <param name="inputSchemaDefinition">Additional parameters for schema mapping between <typeparamref name="TSrc"/> and input data.
+        /// Useful when dealing with annotations.</param>
+        /// <param name="outputSchemaDefinition">Additional parameters for schema mapping between <typeparamref name="TDst"/> and output data.
+        /// Useful when dealing with annotations.</param>
+        public static StatefulCustomMappingEstimator<TSrc, TDst, TState> StatefulCustomMapping<TSrc, TDst, TState>(this TransformsCatalog catalog, Action<TSrc, TDst, TState> mapAction,
+            Action<TState> stateInitAction, string contractName, SchemaDefinition inputSchemaDefinition = null, SchemaDefinition outputSchemaDefinition = null)
```

Comment:
So I guess this is right, but there's something that concerns me about letting the user to change the input and output SchemaDefinition of the transformer.

Some time ago there was this issue where the behavior of CustomMappingTransformer was different between saving the model to disk, and after loading it. My conclusion on the issue was that when saving the model, the information of the SchemaDefinition wasn't saved into disk, and so when loading back the model that information got lost.

https://github.com/dotnet/machinelearning/issues/3988

The issue remains unfixed, but I think it was closed because it's not in our plans to fix it soon and there existed a workaround for that particular issue.

In fact, I am not 100% sure if my conclusion there was correct (so perhaps having input from @yaeldekel would help), but if I was right then I'd think it's better to not let the user manipulate the SchemaDefinitions so that they don't encounter this issue when loading back their models.

On the other hand, I am also not completely sure in which case would the end user require to modify the SchemaDefinition. There's actually only 1 test (and no samples) where the SchemaDefinitions of a CustomMappingTransformer are modified, and I am still not clear about what use case would that test correspond to.

Anyway, it's just a minor concern I have.

Reply:
Interesting. It seems to me that the fix for this would be to serialize both the input and the output `SchemaDefinition`s in `LambdaTransform.SaveCustomTransformer`, and then have the `ICustomMappingFactory.CreateTransformer` method take an optional input/output `SchemaDefinition`. 

Since the issue has been closed (because of low priority I guess?), should I disallow these input parameters in this new class? We cannot remove them from the existing `CustomMappingEstimator` class since that would be a breaking change, but we can at least not introduce the new class with a bug (even if it's just an edge case usage). What do you think? I'm good either way.

---
In reply to: [401514473](https://github.com/dotnet/machinelearning/pull/4676#discussion_r401514473) [](ancestors = 401514473)

Reply:
I'd think it's better not to expose in here the input and output SchemaDefinition to the user. And only expose them if in the future someone requests it or if we actually get to solve the problem of serializing/deserializing SchemaDefinitions.

As for this issue on `CustomMappingTransformer`, I understand that we can't remove the SchemaDefinition parameters from the API. So I will reopen that issue, and tag it as 'breaking-api' so that it gets reviewed in the future if ML.NET 2.0 ever gets planned.